### PR TITLE
Annotate few functions and methods

### PR DESCRIPTION
**Pull request by takeda**
This is related to https://github.com/prometheus/client_python/issues/491, it doesn't resolve it, but it could be a step into that direction.

This theoretically should not break python 2.7

### CODE REVIEW

**Comment by takeda** [comment]
@csmarchbanks is there a chance to get this merged? Should I update this patch to resolve current conflicts?

**Comment by csmarchbanks** [comment]
Hello, sorry, I missed this one! Thank you for the PR. I am hoping to review https://github.com/prometheus/client_python/pull/718 this week, and merge it soon. At that point I believe we can support full typing instead of just the annotations. Would it make sense to wait until after #718 is merged?

**Comment by takeda** [comment]
Yeah, that's fine.

**Comment by takeda** [comment]
@csmarchbanks I rebased my code and adapted it to use annotations as 3.6+ is now required. I currently added types that my code is utilizing. This can be used to show how to add types to other parts of the project.

BTW: since types are not used when the code is run and adding typing involves time, the types could be added incrementally over time.

File: prometheus_client/context_managers.py
Comment (csmarchbanks) [inline prometheus_client/context_managers.py:15]: :thinking: Is there any downside to importing `Counter` all of the time and keeping this type without quotes similar to other cases? To use this function the user will have to import Counter anyway.

File: prometheus_client/metrics.py
Comment (csmarchbanks) [inline prometheus_client/metrics.py:18]: Rather than using `T` could we just use `"MetricWrapperBase"` to make everything clearer later on when reading the functions?

File: prometheus_client/metrics.py
Comment (csmarchbanks) [inline prometheus_client/metrics.py:222]: This no longer needs to be a comment.

File: prometheus_client/samples.py
Comment (csmarchbanks) [inline prometheus_client/samples.py:41]: We should probably use `Dict` here to be consistent with other label sets we use in for example `inc` on a counter. Alternatively we could have a custom type for label sets that is `Dict[str, str]`?

**Review by csmarchbanks — COMMENTED** [review]
Awesome, thanks for the update. Generally this is looking pretty good, I left a handful of questions/comments.

File: prometheus_client/context_managers.py
Comment (csmarchbanks) [reply prometheus_client/context_managers.py:15]: Ohh wait, nevermind. Circular import, resolving this conversation.

File: prometheus_client/metrics.py
Comment (takeda) [reply prometheus_client/metrics.py:18]: That wouldn't work the same way.

This type is used as a generic, it's basically bound to the class and ensures that type returned is of the same type as the mothod's class. Probably I might not even need the bound parameter here.

File: prometheus_client/metrics.py
Comment (takeda) [reply prometheus_client/metrics.py:222]: Ooops, will fix it.

File: prometheus_client/samples.py
Comment (takeda) [reply prometheus_client/samples.py:41]: I can change it to `Dict` I'm not as familiar with the code so I put `Mapping` just in case there's a different type that act like a dict.

Comment (takeda) [reply prometheus_client/samples.py:41]: Do you think that type will be used frequently? If so, maybe it would be warranted to declare it specifically. For now I just changed it to `Dict[]`

File: prometheus_client/metrics.py
Comment (csmarchbanks) [reply prometheus_client/metrics.py:18]: Ahh, yeah that makes sense. Thanks!

File: prometheus_client/samples.py
Comment (csmarchbanks) [reply prometheus_client/samples.py]: I think just a Dict for now makes sense, it would be easiest for users rather than navigating to another type definition.

File: prometheus_client/decorator.py
Comment (csmarchbanks) [inline prometheus_client/decorator.py:42]: This file is actually a copy of http://pypi.python.org/pypi/decorator, you can see the file header for more details. That does mean we should not edit it so as to make future upgrades easier.

**Review by csmarchbanks — COMMENTED** [review]
One small comment that I missed the first time around, otherwise :+1:. Thanks!

File: prometheus_client/decorator.py
Comment (takeda) [reply prometheus_client/decorator.py]: I removed it, but out of curiosity. Why is this needed and why contextmanager is used?

For example:

```python
class ExceptionCounter:
    def __init__(self, counter: "Counter", exception: Type[BaseException]) -> None:
        self._counter = counter
        self._exception = exception

    def __enter__(self) -> None:
        pass

    def __exit__(self, typ: Optional[Type[BaseException]], value: Optional[BaseException], traceback: Optional[TracebackType]) -> Literal[False]:
        if isinstance(value, self._exception):
            self._counter.inc()
        return False

    def __call__(self, f: "F") -> "F":
        def wrapped(func, *args, **kwargs):
            with self:
                return func(*args, **kwargs)

        return decorate(f, wrapped)
```

could be:

```python
@contextmanager
def exception_counter(counter: "Counter", exception: Type[BaseException]) -> None:
  yield
  if isinstance(value, self._exception):
    self._counter.inc()
```

Comment (csmarchbanks) [reply prometheus_client/decorator.py]: You can find the history of the vendored in file in the PR that added it, as well as the linked issue: https://github.com/prometheus/client_python/pull/91.

I think you probably could restructure the various context managers like you suggest now that we do not support older python versions. I believe that before 3.2 just using `@contextmanger` would not allow you to use it as a decorator as well so the class was nicer.

**Review by csmarchbanks — APPROVED** [review]
Thanks!

**Pull request merged**
